rustfmt-core/rustfmt-lib/src/visitor.rs

```diff
@@ -594,67 +597,95 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
+                    self.rewrite_required_fn(indent, ii.ident, sig, &ii.generics, ii.span);
                 self.push_rewrite(ii.span, rewrite);
             }
-            ast::ImplItemKind::OpaqueTy(ref generic_bounds) => {
-                let rewrite = rewrite_opaque_impl_type(
-                    &self.get_context(),
-                    ii.ident,
-                    &ii.generics,
-                    generic_bounds,
-                    self.block_indent,
-                );
+            ast::AssocItemKind::Const(..) => self.visit_static(&StaticParts::from_impl_item(ii)),
+            ast::AssocItemKind::TyAlias(_, ref ty) => {
+                let rewrite_associated = || {
+                    rewrite_associated_impl_type(
+                        ii.ident,
+                        ii.defaultness,
+                        ty.as_ref(),
+                        &ii.generics,
+                        &self.get_context(),
+                        self.block_indent,
+                    )
+                };
+                let rewrite = match ty {
+                    None => rewrite_associated(),
+                    Some(ty) => match ty.kind.opaque_top_hack() {
+                        Some(generic_bounds) => rewrite_opaque_impl_type(
+                            &self.get_context(),
+                            ii.ident,
+                            &ii.generics,
+                            generic_bounds,
+                            self.block_indent,
+                        ),
+                        None => rewrite_associated(),
+                    }
+                };
                 self.push_rewrite(ii.span, rewrite);
             }
-            ast::ImplItemKind::Macro(ref mac) => {
+            ast::AssocItemKind::Macro(ref mac) => {
                 self.visit_mac(mac, Some(ii.ident), MacroPosition::Item);
             }
         }
     }
 
     fn visit_mac(&mut self, mac: &ast::Mac, ident: Option<ast::Ident>, pos: MacroPosition) {
-        skip_out_of_file_lines_range_visitor!(self, mac.span);
+        skip_out_of_file_lines_range_visitor!(self, mac.span());
 
         // 1 = ;
         let shape = self.shape().saturating_sub_width(1);
         let rewrite = self.with_context(|ctx| rewrite_macro(mac, ident, ctx, shape, pos));
-        self.push_rewrite(mac.span, rewrite);
+        let (span, rewrite) = match macro_style(mac, &self.get_context()) {
+            DelimToken::Bracket | DelimToken::Paren if MacroPosition::Item == pos => {
+                let search_span = mk_sp(mac.span().hi(), self.snippet_provider.end_pos());
+                let hi = self.snippet_provider.span_before(search_span, ";");
+                let target_span = mk_sp(mac.span().lo(), hi + BytePos(1));
+                let rewrite = rewrite.map(|rw| {
+                    if !rw.ends_with(";") {
+                        format!("{};", rw)
+                    } else {
+                        rw
+                    }
+                });
```

Comment:
This was done to fix two snippets that were failing in the tests due to the `;` being stripped. I think that was related to the same span change around missing semis so correcting it here. 

The two failing snippets without this were:
https://github.com/rust-lang/rustfmt/blob/facba6a58a50fbdb1592748bb25defd95b5322c4/rustfmt-core/rustfmt-lib/tests/target/macros.rs#L239-L242

and 
https://github.com/rust-lang/rustfmt/blob/facba6a58a50fbdb1592748bb25defd95b5322c4/rustfmt-core/rustfmt-lib/tests/target/issue-2916.rs#L1-L2

Reply:
Note that on master/rustfmt 1.x that any whitespace between the closing delim and semi is not currently removed for these two snippets, so I think this actually fixes a very minor existing issue. 

For example with the below/linked snippet, the whitespace between the delim and semi remains after formatting with 1.x/master:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=649ebcbc83ba416d42392084d91ba266
```rust
gfx_pipeline!(pipe { 
     vbuf: gfx::VertexBuffer<Vertex> = (), 
     out: gfx::RenderTarget<ColorFormat> = "Target0", 
 })       ;
```

I enhanced those two test snippets to ensure that rustfmt removes that extra whitespace